lib/Sema/CSApply.cpp

```diff
@@ -1950,6 +1950,33 @@ namespace {
       tc.validateDecl(MaxIntegerTypeDecl);
       auto maxType = MaxIntegerTypeDecl->getUnderlyingTypeLoc().getType();
 
+      // Make sure that the literal value will fit within the bit width of the
+      // 'maxType'. Otherwise emit diagnositcs.
+      if (IntegerLiteralExpr *intLitExpr = dyn_cast<IntegerLiteralExpr>(expr)) {
+        unsigned maxWidth =
+                    maxType->castTo<BuiltinIntegerType>()->getGreatestWidth();
+        APInt magnitude = intLitExpr->getRawMagnitude();
+        unsigned magWidth = magnitude.getActiveBits();
+        bool isNegative = intLitExpr->isNegative();
+
+        // Compute the literal bit width in the signed two's complement form.
+        // This is generally one more than the magnitude width, but is the same
+        // when the literal is of the form -2^i, for some i.
+        unsigned signedLitWidth =
+          (isNegative && (magnitude.countTrailingZeros() == magWidth - 1)) ?
+          magWidth : (magWidth + 1);
+
+        if (signedLitWidth > maxWidth) { // overflow?
+          SmallString<10> litStr;
+          if (intLitExpr->isNegative())
+            litStr += '-';
+          litStr += intLitExpr->getDigitsText();
+          tc.diagnose(expr->getLoc(), diag::integer_literal_overflows_maxwidth,
+                      maxWidth, signedLitWidth, litStr);
+          return nullptr;
```

Comment:
We are trying to reduce the number of `return nullptr` from CSApply because it means that constraint system shouldn't have formed solution like that at the first place. It seems that this check would be a better fit for `CSGen#visitLiteralExpr` because it uses raw data from the expression anyway, WDYT?

Reply:
@xdein Thanks for the comments. I would like to add it CSGen. However,  'MaxIntegerTypeDecl'  seems to be materialized only in CSApply just before the point where I have added this code (line 1933). I am wondering if it possible to know the maximum bit width of an integer literal in CSGen#visitLiteralExpr.

Reply:
@ravikandhadai It doesn't seem like it should be a problem to move the logic from 1933 - 1951 over to CSGen from here because it just does lookup into stdlib? It looks like nothing else in CSApply is using `MaxIntegerTypeDecl` anyway.

Reply:
@CodaFi It is possible but doesn't seem ideal. In the SIL generated by SILGen phase, integer literals get truncated to 2048 bits. So we don't see the original literal text unless we walk through the AST.  This also delays diagnostics (we don't need constant folding machinery to do this check.) Also, I guess handling Enum cases at the level of SIL opts may result in unnecessary overheads.

Reply:
You can walk back along Source Locs to AST nodes.  The literal instructions should support that.  @xedin is right here: failing the solution set this late for this kind of diagnostic isn’t great.

Reply:
@codaFi  I will try adding it to CSGen as suggested by @xedin.